### ISSUE TITLE
WSL 2.0: `networkingMode=mirrored` makes Docker unable to forward ports

### ISSUE THREAD

**Issue by halimsamy** [microsoft/WSL#10494]
### Windows Version

Microsoft Windows [Version 10.0.22631.2338]

### WSL Version

2.0.0.0

### Are you using WSL 1 or WSL 2?

- [X] WSL 2

### Kernel Version

5.15.123.1-1

### Distro Version

Ubuntu 20.04.6 

### Other Software

Docker version 24.0.6, build ed223bc

### Repro Steps

- Change `networkingMode` to `mirrored`
- `docker run -d -p 8080:80 nginx:alpine` (example)
- Go to `localhost:8080`
- Cannot connect
- Try from WSL `curl http//localhost:8080` same issue

### Expected Behavior

To forward the port and be able to connect to my containers

### Actual Behavior

Doesn't forward the port, so I cannot connect to my containers.

**Comment by driver1998** [comment]
Interestingly, the port forwarding does work from another machine on the same network as host. Just not on the host machine itself.

Machine A:
- Change networkingMode to mirrored
- docker run -d -p 8080:80 nginx:alpine
- Go to localhost:8080
- Cannot connect
- Set firewall rules to allow 8080 inbound
- Go to localhost:8080
- Cannot connect

Machine B:
- Go to A:8080
- works

**Comment by greenhat616** [comment]
Same issue

**Comment by giovannicandido** [comment]
Same problem here

**Comment by over-star** [comment]
Same problem here

**Comment by speller** [comment]
In my case, Docker ports are not exposed to the host Windows machine at all after the upgrade to 2.0 even if I set `networkingMode=NAT` which is the default value. I can't see opened ports in the `tcpview.exe` utility. Previously, I was able to see them opened by the `wslhost.exe` process (or similar name).

**Comment by Yukari316** [comment]
same here

**Comment by tuxiaobei233** [comment]
+1

**Comment by bkfino** [comment]
+1

**Comment by ptr1120** [comment]
+1

**Comment by halimsamy** [comment]
@benhillis Any ideas/updates here?

**Comment by halimsamy** [comment]
@benhillis Any ideas/updates here?

**Comment by DukeCrimson** [comment]
same here

**Comment by tumugin** [comment]
+1

**Comment by Aerglonus** [comment]
Same issue here, the only way I found to make it "work" was adding ignoredPorts=8080, but if container has something like 4800:8080 need to add both ports to `ignoredPorts=8080,4800` still doing this some containers might not work at all, also if cloudflare tunnel is running on WSL this will break to when mirrored and theres no way to know which ports is using to add it to the `ignoredPorts`

**Comment by halimsamy** [comment]
> Same issue here, the only way I found to make it "work" was adding ignoredPorts=8080 to the wslconfig, but if container has something like 4800:8080 need to add both ports to `ignoredPorts=8080,4800` still doing this some containers might not work at all, also if cloudflare tunnel is running on WSL this will break to when mirrored and theres no way to know which ports is using to add it to the `ignoredPorts`

I think that's a workaround that would work... but I am waiting for someone from the WSL team to answer us here, did they get to know that is the issue, and if they have any plans to fix this?

**Comment by lengthofrope** [comment]
> Same issue here, the only way I found to make it "work" was adding ignoredPorts=8080 to the wslconfig, but if container has something like 4800:8080 need to add both ports to `ignoredPorts=8080,4800` still doing this some containers might not work at all, also if cloudflare tunnel is running on WSL this will break to when mirrored and theres no way to know which ports is using to add it to the `ignoredPorts`

That seems to work. Nice find.

**Comment by zcobol** [comment]
`ignorePorts` is misleading. What it does is making in this case port `8080` available to request inside the container only, but is not forwarded to Windows side. Look for more details at https://learn.microsoft.com/en-us/windows/wsl/wsl-config

If you run  `get-nettcpConnection -LocalPort 8080` there's nothing listening on that port.

After starting the web server with `docker run -d -p 8080:80 nginx:alpine` requests inside the WSL container works:

```
zcobol@debian:~$ curl -I localhost:8080
HTTP/1.1 200 OK
Server: nginx/1.25.2
Date: Mon, 25 Sep 2023 16:41:36 GMT
Content-Type: text/html
Content-Length: 615
Last-Modified: Tue, 15 Aug 2023 19:24:07 GMT
Connection: keep-alive
ETag: "64dbd0d7-267"
Accept-Ranges: bytes

```
Requests from Windows side timeout with:

`curl: (28) Failed to connect to localhost port 8080 after 21276 ms: Couldn't connect to server`

From another server it work, as mentioned by @driver1998 :

```
tux@raspi:~$ curl -I 192.168.1.105:8080
HTTP/1.1 200 OK
Server: nginx/1.25.2
Date: Mon, 25 Sep 2023 16:43:45 GMT
Content-Type: text/html
Content-Length: 615
Last-Modified: Tue, 15 Aug 2023 19:24:07 GMT
Connection: keep-alive
ETag: "64dbd0d7-267"
Accept-Ranges: bytes
```

**Comment by mew1033** [comment]
Just ran into this as well. AWS `sam build --use-container` doesn't work. Gives the same error
```
 Ports are not available: exposing port TCP 127.0.0.1:5232 -> 0.0.0.0:0: listen tcp 127.0.0.1:5232: bind: Only one usage of each socket address (protocol/network address/port) is normally permitted.
```
(The port number changes randomly every time)

**Comment by halimsamy** [comment]
> Just ran into this as well. AWS `sam build --use-container` doesn't work. Gives the same error
> 
> ```
>  Ports are not available: exposing port TCP 127.0.0.1:5232 -> 0.0.0.0:0: listen tcp 127.0.0.1:5232: bind: Only one usage of each socket address (protocol/network address/port) is normally permitted.
> ```
> 
> (The port number changes randomly every time)
> 
> Based on this: https://github.com/aws/aws-sam-cli/blob/c5b9b1e399a1e5c938ef72934a14ede934e17bac/samcli/local/docker/container.py#L124-L125 I added every port from 5000-9000 to the `ignoredPorts` list. It's ugly, but for now it works.

It seems like it's generally a problem with WSL itself. I am looking for a fix soon, since the new network mode is so much useful but it's not useable (usefully) in the current state.

**Comment by imacarthur741** [comment]
I ma having the same issue with apache. No changes other than added --experimental but now nothing works.

(98)Address already in use: AH00072: make_sock: could not bind to address [::]:80
(98)Address already in use: AH00072: make_sock: could not bind to address 0.0.0.0:80
no listening sockets available, shutting down
[no listening sockets available, shutting down](apache2.service: Control process exited, code=exited, status=1/FAILURE)

root@ACER-Nitro:/usr/sbin# lsof -nP -iTCP -sTCP:LISTEN
COMMAND PID USER FD TYPE DEVICE SIZE/OFF NODE NAME
systemd-r 308 systemd-resolve 14u IPv4 22507 0t0 TCP 127.0.0.53:53 (LISTEN)
mysqld 385 mysql 21u IPv4 32886 0t0 TCP 127.0.0.1:33060 (LISTEN)
mysqld 385 mysql 23u IPv4 27396 0t0 TCP 127.0.0.1:3306 (LISTEN)
miniserv. 1867 root 5u IPv4 52863 0t0 TCP *:10000 (LISTEN)

**Comment by zohaibhassan156** [comment]
Same problem

**Comment by zohaibhassan156** [comment]
`ignorePorts` don't resolve this issue like in my mailpit can't access from windows side with `ignoredPorts=8025,1025`
```
docker run -d \
--name=mailpit \
--restart unless-stopped \
-p 8025:8025 \
-p 1025:1025 \
axllent/mailpit
```

**Comment by zohaibhassan156** [comment]
`ignoredPorts=8025,1025` work for me too. I had to use `mailpit` in `docker`
```
docker run -d \
--name=mailpit \
--restart unless-stopped \
-p 8025:8025 \
-p 1025:1025 \
axllent/mailpit
```

**Comment by shigenobuokamoto** [comment]
there seem to be two issues why Docker containers cannot connect from Windows.

1. `ignorePorts` works
Docker Desktop, probably

2. `the port forwarding does work from another machine on the same network as host. Just not on the host machine itself.`
docker-ce that installed on Linux

temporary measures for `2` ....
`disable iptables and use docker-proxy`

/etc/docker/daemon.json
```
{
    "iptables": false
}
```


when using mirrored, the behavior seems to be different from the previous localhostforwarding.


use docker-proxy(listen on Linux)
```
windows  --->  linux(WSL)
  localhostforwarding (until)
ok    127.0.0.1   --->     127.0.0.1(lo)
ok    127.0.0.1   <---     127.0.0.1(lo)

  netowrkingMode=mirrored
ok    127.0.0.1   --->     127.0.0.1(loopback0)
ok    127.0.0.1   <---     127.0.0.1(loopback0)
```

interface is different, but the behavior remains the same.

use iptables(listeon on container)
```
windows  --->  linux(WSL) ---> Docker container
  localhostforwarding (until)
ok    127.0.0.1   ---> 127.0.0.1(lo)    /   172.x.x.1(br-xxx)  --->  172.x.x.x(eth0)
ok    127.0.0.1   <--- 127.0.0.1(lo)    /   172.x.x.1(br-xxx)  <---  172.x.x.x(eth0)

  netowrkingMode=mirrored
ok    127.0.0.1   ---> 127.0.0.1(loopback0)/127.0.0.1(br-xxxx) --->  172.x.x.x(eth0)
                                          ~~~~~~~~~
ng                                                           <- - -  172.x.x.x(eth0)
```

via localhostforwarding(until), source address(Windows) was the docker network gateway (=pointing to linux).

via mirrored, source address is 127.0.0.1.
for this reason that packet is returned to 127.0.0.1(localhost on container) and does not reach Windows.
in the case of access from another node(PC), there is no problem because source is his address.

**Comment by i-curve** [comment]
Same problem

**Comment by jogiji** [comment]
Same problem :) was on the verge of reinstalling the whole system..

**Comment by rozbo** [comment]
Any update?

**Comment by halimsamy** [comment]
`ignorePorts`  (with `iptables` disabled) seems to work but it doesn't allow you to connect to the container from the Windows Machine. However, it doesn't work with Redis (running inside docker for some reason, so I didn't have enough time to look it up).

**Comment by halimsamy** [comment]
`ignorePorts`  (with `iptables` disabled) seems to work but it doesn't allow you to connect to the container from the Windows Machine. However, it doesn't work with Redis (running inside docker for some reason, so I didn't have enough time to look it up).

**Comment by NasdaqScott** [comment]
+1 for priority. There is a large contingent of corporate devs who need docker and mirrored mode (vpn/sec reqs). 

ignoredPorts really isn't an ideal solution. Hopefully we can get an official update on this item as it's already quite popular.

**Comment by rasta-mouse** [comment]
Came across this issue today when starting my RabbitMQ container.  Adding my +1 for visibility.

**Comment by fanthos** [comment]
I tested with disable iptables for docker in `/etc/docker/daemon.json` and use `docker-proxy` instead. It works on my device.

**Comment by DukeCrimson** [comment]
Using ignorePorts does not work on my laptop.
I only installed docker in my wsl2 and didn't use docker desktop client in Windows.
Is that the reason?

**Comment by caaat233** [comment]
> 使用ignorePorts 在我的笔记本电脑上不起作用。 我的意思是我仍然无法从 Windows 访问 docker 服务端口。 我只在我的wsl2中安装了docker，并没有在Windows中使用docker桌面客户端。 是这个原因吗？

添加下面配置到wslconfig

```
[experimental]
autoMemoryReclaim=gradual
networkingMode=mirrored
dnsTunneling=true
firewall=false
autoProxy=false
hostAddressLoopback=true
```

并且添加 `"iptables": false` 到 `/etc/docker/daemon.json` 就好了

**Comment by felipecrs** [comment]
@caaat233 it works indeed, thanks a lot! All it takes is to put `{"iptables": false}`.

**Comment by keith-horton** [comment]
Thanks for the work around. We are engaged with the Docker team to have better integration.

**Comment by NasdaqScott** [comment]
@keith-horton Why is this being closed when you are still actively working with Docker? Is there another issue that I can track? Is this workaround an official workaround suggested by MS? :) Are there any negative/downsides from turning iptables off in Docker w/ WSL?

**Comment by keith-horton** [comment]
Sorry, it was close for internal tracking of issues.
If it's more appropriate, we can keep this open as well.

**Comment by keith-horton** [comment]
" Are there any negative/downsides from turning iptables off in Docker w/ WSL?"

Sorry, I don't know Docker well enough to explain their various options. Do they have a forum to ask this?

**Comment by mew1033** [comment]
> Sorry, it was close for internal tracking of issues. If it's more appropriate, we can keep this open as well.

Does that mean a fix has been developed and we're just waiting on a new version? If so, awesome! Thank you!

**Comment by keith-horton** [comment]
Sorry, we are working with the Docker developers so their solution works well with WSL. We don't yet have a fix from our side.

**Comment by halimsamy** [comment]
> Interestingly, the port forwarding does work from another machine on the same network as host. Just not on the host machine itself.
> 
> Machine A:
> 
> * Change networkingMode to mirrored
> * docker run -d -p 8080:80 nginx:alpine
> * Go to localhost:8080
> * Cannot connect
> * Set Windows firewall rules to allow 8080 inbound
> * Go to localhost:8080
> * Cannot connect
> 
> Machine B:
> 
> * Go to A:8080
> * works

After removing Docker Desktop and installing Docker directly on WSL, I got the containers to start.

However, Containers that expose ports are not accessible from the Windows machine or any device on the network
Yet, If you started a Node server for example, everything works like charm and can be accessed from Windows and any device on the network

Meanwhile `lsof` shows the following:

```
➜  ~ sudo lsof -i:8080
COMMAND     PID USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
docker-pr 13090 root    4u  IPv4 189364      0t0  TCP *:http-alt (LISTEN)
docker-pr 13100 root    4u  IPv6 177996      0t0  TCP *:http-alt (LISTEN)
➜  ~ sudo lsof -i:3001
COMMAND  PID  USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
node    3246 halim   26u  IPv6  21934      0t0  TCP *:3001 (LISTEN)
```

Not sure what does `*:http-alt` means but I think it's related to the problem!

**Comment by BlueTree242** [comment]
Disabling iptables and using docker-proxy is a bad idea after all as docker-proxy has overhead

**Comment by panoshu** [comment]
> > 使用ignorePorts 在我的笔记本电脑上不起作用。 我的意思是我仍然无法从 Windows 访问 docker 服务端口。 我只在我的wsl2中安装了docker，并没有在Windows中使用docker桌面客户端。 是这个原因吗？
> 
> 添加下面配置到wslconfig
> 
> ```
> [experimental]
> autoMemoryReclaim=gradual
> networkingMode=mirrored
> dnsTunneling=true
> firewall=false
> autoProxy=false
> hostAddressLoopback=true
> ```
> 
> 并且添加 `"iptables": false` 到 `/etc/docker/daemon.json` 就好了

it works for me

**Comment by mendhak** [comment]
So the two workarounds I see so far, both have their negatives but worth listing anyway. Correct any of my misunderstandings. 

**Option1 - without docker desktop**

Edit `%USERPROFILE%\.wslconfig` file (create it, if doesn't exist) and add this section

```
[experimental]
networkingMode=mirrored
hostAddressLoopback=true
```

Restart WSL, then in your distro, edit `/etc/docker/daemon.json` and add

```json
{
"iptables": false
}
```

Then `sudo systemctl restart docker`. 

With this, both normal port usage like `python3 -m http.server` and docker port usage like `docker run -p "8080:8080" --rm -t mendhak/http-https-echo:26` are accessible from Windows side. The disadvantage is that iptables is now disabled. 

**Option 2 - with Docker Desktop**

Edit `%USERPROFILE%\.wslconfig` file (create it, if doesn't exist) and add this section

```
[experimental]
networkingMode=mirrored
hostAddressLoopback=true
ignoredPorts = 8000,8080
```

Restart WSL and ensure Docker Desktop is running. With this, the normal port usage stops working if it uses a port listed above. eg the Python http.server isn't accessible over 8000 anymore. Instead use a different port, like `python3 -m http.server 8001`. Docker ports will work as normal as long as the port is listed in the `ignoredPorts` above, like `docker run -p "8080:8080" --rm -t mendhak/http-https-echo:26`

The disadvantage here is you have to know which ports you'll be using with Docker. And it doesn't look like `ignoredPorts` accepts ranges of ports either so it can get pretty tedious.

**Comment by AzureZeng** [comment]
Docker Desktop 4.26.0 Release Notes
[For Windows](https://docs.docker.com/desktop/release-notes/#for-windows)
\* **Added support for WSL mirrored mode networking (requires WSL v2.0.4 and up).**
\* Added missing signatures on DLL and VBS files.

**Comment by mendhak** [comment]
Thank you for posting @AzureZeng 

Would appreciate a few more people checking to confirm, but I just did a quick test and it seems to be working better than before, hopefully no other issues appear 🤞 

In `%USERPROFILE%\.wslconfig` I only put this and nothing else:

```
[experimental]
networkingMode=mirrored
```

and updated to Docker Desktop 4.26.0

I am now able to

```
python3 -m http.server
```

and also 

```
docker run -p 8080:8080 --rm -t mendhak/http-https-echo:31
```

Both are visible from the Windows side in a browser.

**Comment by felipecrs** [comment]
Ok, that's cool, but I hope this issue does not get closed already. What about people who are not using Docker Desktop? (which is not free and neither open source)

**Comment by robpomeroy** [comment]
Re the Docker update to 4.26.0 - WSL networking is working correctly for me in bridged mode. My Docker containers have stopped binding to my preferred NIC's IP though. Ports are exposed on localhost, but not my LAN IP. That will require some investigation and potentially, reconfiguration.

**Comment by AzureZeng** [comment]
> Ok, that's cool, but I hope this issue does not get closed already. What about people who are not using Docker Desktop? (which is not free and neither open source)

See above, set `networkingMode=mirrored' and `hostAddressLoopback=true` in `.wslconfig`, then set `"iptables": false` in `/etc/docker/daemon.json`.

I think that there will not have more convenient solution than this, unless Docker team want to improve the experience that running Docker in WSL (without Docker Desktop).

**Comment by felipecrs** [comment]
Yes, I'm aware of that and that's what I am using. BTW, I don't need to set `hostAddressLoopback=true` and apparently it still works fine.

But I do expect some answer from either the WSL team or Docker as to if this is the recommended/final approach or not.

**Comment by npuichigo** [comment]
> ###
> Docker version 24.0.6, build ed223bc

I always find the solution right after it's been fixed. So lucky.

**Comment by LordHumphrey** [comment]
> > Docker version 24.0.6, build ed223bc
> 
> I always find the solution right after it's been fixed. So lucky.

What's your solution? Could you share with me your configuration？

**Comment by npuichigo** [comment]
@LordHumphrey Update my quote. Just use docker desktop 4.26.0

**Comment by craigloewen-msft** [comment]
I'm closing this issue out as it's now fixed! Please upgrade to docker desktop 4.26.0 to get the fix. Thank you all! :)

**Comment by felipecrs** [comment]
@craigloewen-msft, any words for the people who are not using Docker Desktop (which is not even free)? Is `echo '{"iptables": false}' > /etc/docker/daemon.json` the proper approach?

**Comment by BlueTree242** [comment]
This issue should not be closed, the issue is still there if you're not using docker desktop. Iptables is more stable then userland proxy

**Comment by shigenobuokamoto** [comment]
even if iptables:true, unless you explicitly set userland-proxy: false, access from localhost will go through userlandproxy.

workaround if you don't want to set iptables:false
https://gist.github.com/shigenobuokamoto/b565d468541fc8be7d7d76a0434496a0

**Comment by craigloewen-msft** [comment]
I think I'm not fully understanding the scenario. Here are the repro steps I tested in both docker-ce (So running as a Linux service in my WSL distro) and on Docker Desktop:

- Change networkingMode to mirrored
- `docker run -d -p 8080:80 nginx:alpine` (example)
- Go to `localhost:8080` in Ede browser
- Test to see if it's succeeding

This worked for both Docker Desktop and docker-ce when using mirrored mode. If you're still seeing this issue could you give me the repro steps where it's coming up for you?

**Comment by felipecrs** [comment]
That's the exact test scenario I have, and it's not working here:


https://github.com/microsoft/WSL/assets/29582865/8f89f3f5-4472-4cc4-a686-a909fedd5985

**Comment by MeikTranel** [comment]
Im running docker `24.0.7` with docker desktop version `4.26.1` and i'm still experiencing the issue:

```
WSL version: 2.0.0.0
Kernel version: 5.15.123.1-1
WSLg version: 1.0.57
MSRDC version: 1.2.4485
Direct3D version: 1.608.2-61064218
DXCore version: 10.0.25880.1000-230602-1350.main
Windows version: 10.0.22631.2861
```


https://github.com/microsoft/WSL/assets/6466560/cb659445-dc55-45a9-ac1a-d9b4388bf8bd

Not sure if i'm doing something wrong but it feels very much not solved 😂 happy to help diagnose ;)

**Comment by thijsputman** [comment]
> * Change networkingMode to mirrored
> * `docker run -d -p 8080:80 nginx:alpine` (example)
> * Go to `localhost:8080` in Ede browser
> * Test to see if it's succeeding

Doing exactly the above (with Firefox instead of Edge 😉) results in a "The connection has timed out"-message.

It works fine if I _either_ set `"iptables": false` _or_ if I add `--network host` to the command (and then connect to `localhost:80` instead of `8080`).

Running `docker-ce/jammy,now 5:24.0.7-1~ubuntu.22.04~jammy amd64` on Ubuntu 22.04 with kernel `5.15.133.1-microsoft-standard-WSL2`.

**Comment by craigloewen-msft** [comment]
I'll reopen this then! For all the folks still seeing this, please collect WSL logs ( instructions here: http://aka.ms/wsllogs ) and then run these repro steps: 

- Change networkingMode to mirrored
- docker run -d -p 8080:80 nginx:alpine (example)
- Go to localhost:8080 in Ede browser
- Test to see if it's succeeding

Then please attach the logs in a GitHub comment on this thread. Lastly *please specify whether you are using Docker Desktop or docker-ce*! Thank you!

**Comment by felipecrs** [comment]
Here are my logs.

I'm using docker-ce 24.0.7, docker desktop is not even installed.

[WslLogs-2023-12-21_17-09-50.zip](https://github.com/microsoft/WSL/files/13745895/WslLogs-2023-12-21_17-09-50.zip)

**Comment by MeikTranel** [comment]
Using Docker Desktop 4.26.1 / reproduced using @craigloewen-msft's provided steps:

[WslLogs-2023-12-22_09-54-49.zip](https://github.com/microsoft/WSL/files/13750765/WslLogs-2023-12-22_09-54-49.zip)

Happy holidays tho 🤗

**Comment by robpomeroy** [comment]
One Patch Tuesday later and it's now working for me (using Docker Desktop 4.26.0). 🤔 Windows 11 23H2 (build 22635.2771).

```
λ wsl --version
WSL version: 2.0.14.0
Kernel version: 5.15.133.1-1
WSLg version: 1.0.59
MSRDC version: 1.2.4677
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.22635.2771
```

**Comment by felipecrs** [comment]
I also re-tested docker-ce, but unfortunately I have the same result still.

**Comment by keith-horton** [comment]
@felipecrs , thank you for the trace. It looks like the wsl service is crashing shortly after startup. Can you follow these steps so that the wslservice.exe crash dmp is collected locally?

https://learn.microsoft.com/en-us/windows/win32/wer/collecting-user-mode-dumps?redirectedfrom=MSDN
(and stack overflow questions if you hit issues: https://stackoverflow.com/questions/30457570/windows-does-not-produce-full-crash-dumps-even-though-registry-is-set-correctly)

Thank you!

**Comment by keith-horton** [comment]
@MeikTranel , thanks for your trace. This is failing to reserve the TCP port for 0.0.0.0:80 in the Linux container. Is that port already used on the Windows side? You can check it via Powershell.

e.g. this shows I have one listener on the IPv6 ephemeral address on port 80.

PS C:\Users\khorton> Get-NetTCPConnection -LocalPort 80

LocalAddress                        LocalPort RemoteAddress                       RemotePort State       AppliedSetting
------------                        --------- -------------                       ---------- -----       --------------
::                                  80        ::                                  0          Listen


We have updated logging that is specific to Networking: https://github.com/microsoft/WSL/blob/master/diagnostics/collect-networking-logs.ps1

If you could stop WSL (wsl --shutdown), start the trace and repro, that would be very helpful.

Thanks!

**Comment by felipecrs** [comment]
However, I capture network logs as you mentioned:

[WslNetworkingLogs-2024-01-03_10-40-57.zip](https://github.com/microsoft/WSL/files/13819580/WslNetworkingLogs-2024-01-03_10-40-57.zip)

**Comment by MeikTranel** [comment]
> @MeikTranel , thanks for your trace. This is failing to reserve the TCP port for 0.0.0.0:80 in the Linux container. Is that port already used on the Windows side? You can check it via Powershell.

The output is the same for me - i have process id 4 (aka the "system" process) listening on IPv6 :: Port 80 but you can see in the video i uploaded before that i switch between mirrored and not mirrored and the binding error inside `httpd` only happens after switching to mirrored networking here: https://github.com/microsoft/WSL/issues/10494#issuecomment-1864202014

I collected another set of logs with the updated scripts of yours - a note i'd like to make is that i changed the exposed ports in my compose to use ports that don't show up under Get-NetTCPConnection and it seems the failed binding happens within the container - its failing to bind 80 from within - windows ports do not seem to matter at that point. I'd understand that it'd fail to bind on the host port but the containerport should not be affected by the hostsystem having IIS running or something in a way that it doesnt even let me bind a non-standard host port - that would make working with other 3rdParty container images very awkward.

[WslNetworkingLogs-2024-01-08_10-31-32.zip](https://github.com/microsoft/WSL/files/13858313/WslNetworkingLogs-2024-01-08_10-31-32.zip)

**Comment by Incipe-win** [comment]
Here are my logs.

I'm using docker-ce 24.0.7
[WslLogs-2024-01-17_10-58-42.zip](https://github.com/microsoft/WSL/files/13958090/WslLogs-2024-01-17_10-58-42.zip)

**Comment by bayeslearner** [comment]
Seems to work fine for me.  
Nothing special has to be done. 
Except according #11034, one of the settings "hostAddressLoopback = true" shouldn't be enabled.

**Comment by felipecrs** [comment]
I tested again, with latest Win11 and WSL. Problem remains the same here.

@bayeslearner are you using mirrored networkingMode? Which distro? iptables enabled in docker daemon?

**Comment by bayeslearner** [comment]
> I tested again, with latest Win11 and WSL. Problem remains the same here.
> 
> @bayeslearner are you using mirrored networkingMode? Which distro? iptables enabled in docker daemon?

Yes, mirrored networking mode.  Ubuntu 22.04.  I'm using docker-desktop , not docker that can also be directly installed on Linux.

**Comment by felipecrs** [comment]
I see. And it was already known that docker desktop was working. :)

**Comment by JaylonGu** [comment]
After run the following command in PowerShell window with admin privileges to [Configure Hyper-V firewall](https://learn.microsoft.com/en-us/windows/security/operating-system-security/network-security/windows-firewall/hyper-v-firewall) settings to allow inbound connections: 

`Set-NetFirewallHyperVVMSetting -Name ‘{40E0AC32-46A5-438A-A0B2-2B479E8F2E90}’ -DefaultInboundAction Allow`

I can not only connect to the container from Windows using localhost, but also from other computer in the same lan using the ip address of the Windows.

Reference：
[mirrored-mode-networking](https://learn.microsoft.com/en-us/windows/wsl/networking#mirrored-mode-networking)

**Comment by xlccccc** [comment]
When I set networkMode=mirrored, I found that the service on wsl can be accessed by windows through 127.0.0.1, but cannot be accessed through LAN ip. At this time, docker forwarding port is normal, but docker cannot be accessed through LAN ip. to services on wsl. So I set hostAddressLoopback=true. At this time, Windows can access all services on WSL through LAN IP, but the docker port cannot be forwarded, and WSL services cannot be accessed through LAN IP. In both cases, docker can ping the LAN IP.
I use docker desktop 4.26.1

**Comment by keith-horton** [comment]
Hi there. By default, you can only communicate to/from the host and container with 127.0.0.1. But if you set [experimental] hostAddressLoopback = true, then you can also use the host IP address.  https://learn.microsoft.com/en-us/windows/wsl/wsl-config#experimental-settings

**Comment by wswind** [comment]
None of these workaround works.   Juse change back to NAT and waiting for fix.

**Comment by felipecrs** [comment]
Just to update, this issue is still present with WSL `2.1.1.0` and docker `25.0.2`.

**Comment by xlccccc** [comment]
After updating, the docker container still cannot access services open on wsl.

**Comment by juicyslew** [comment]
This is still an issue.

**Comment by HITCJK** [comment]
> temporary measures for ....`2` `disable iptables and use docker-proxy`
> 
> /etc/docker/daemon.json
> 
> ```
> {
>     "iptables": false
> }
> ```
it works for me,but this causes me to not be able to use iptables and I'm still waiting for a real fix. Also, I've heard that Docker Desktop 4.26 and later adds support for mirrored network mode.

**Comment by foureight84** [comment]
> > temporary measures for ....`2` `disable iptables and use docker-proxy`
> > /etc/docker/daemon.json
> > ```
> > {
> >     "iptables": false
> > }
> > ```
> 
> it works for me,but this causes me to not be able to use iptables and I'm still waiting for a real fix. Also, I've heard that Docker Desktop 4.26 and later adds support for mirrored network mode.

docker iptables: false confirmed working.

For anyone trying to connect services in WSL2 to connect to Android Emulator running in windows and have access to docker containers, this workaround works. Make sure to unset `ADB_SERVER_SOCKET` if set previously in WSL2 guest.

**Comment by LeorGreenberger** [comment]
> https://gist.github.com/shigenobuokamoto/b565d468541fc8be7d7d76a0434496a0
> 
> how about this?

This worked!!

**Comment by shigenobuokamoto** [comment]
@LeorGreenberger 
what distribution are you using?

**Comment by LeorGreenberger** [comment]
> @LeorGreenberger what distribution are you using?

Debian 12

**Comment by shigenobuokamoto** [comment]
@LeorGreenberger 
i cannot reproduce it.
however, there were some parts where the escape was missing, so I fixed it.
could you please try this?

https://gist.github.com/shigenobuokamoto/b565d468541fc8be7d7d76a0434496a0

**Comment by LeorGreenberger** [comment]
@shigenobuokamoto  This works! I reloaded, stopped and started a bunch of times and I don't see any escape errors. On stopping the service, I get the following warning, but I am guessing this is expected:

```
Mar 07 13:57:29 REMLT28981 systemd[1]: Stopping network-mirrored.service...
Mar 07 13:57:29 REMLT28981 sh[221292]: # Warning: table ip nat is managed by iptables-nft, do not touch!
Mar 07 13:57:29 REMLT28981 systemd[1]: network-mirrored.service: Deactivated successfully.
Mar 07 13:57:29 REMLT28981 systemd[1]: Stopped network-mirrored.service.
```

**Comment by andreaprilucm** [comment]
@shigenobuokamoto Your fix works!!! I ran your service start command as root at the command line. Not tested the stop service yet. Our use case is a windows fat client connected to a backend  composed of a bunch of containers started with docker compose in WSL2.

You saved my life!

**Comment by felipecrs** [comment]
> You saved my life!

I wonder why `iptables: false` would not be enough for you.

**Comment by shigenobuokamoto** [comment]
@felipecrs 
> I wonder why iptables: false would not be enough for you.

most issues can be resolved by modifying daemon.json. however, the usability may change by not using iptables.
bigger issues is that cannot use docker unless setting daemon.json.
my script is just a sample implementation workaround, but if it has a reasonable effect without any negative effects on everyone, i hope that i will be able to propose its inclusion in `/init`.

https://gist.github.com/shigenobuokamoto/b565d468541fc8be7d7d76a0434496a0


my script is very simple, just apply the following rules to nftables.

```
table ip nat {
    chain WSLPREROUTING {
        type nat hook prerouting priority dstnat - 1; policy accept;
        iif "loopback0" ip daddr 127.0.0.1 counter dnat to 127.0.0.1;
    }
}
```
this chain is processed immediately before PREROUTING, and will DNAT destination of the packet that arrives from loopback0 interface to 127.0.0.1 to 127.0.0.1. it may seem like a wasteful process, but adding this action breaks the prerouting hook and disables any Docker rules set in the PREROUTING chain.

**using userland-proxy only for access from localhost is the same as Docker's default behavior.**

to: who have already used it
i have made some changes from the previous one,
```
before
  add chain   ip nat PREROUTING { type nat hook prerouting priority dstnat; policy accept; };
  insert rule ip nat PREROUTING iif loopback0 counter accept comment mirrored;

after
  add chain   ip nat WSLPREROUTING { type nat hook prerouting priority dstnat - 1; policy accept; };
  insert rule ip nat WSLPREROUTING iif loopback0  ip daddr 127.0.0.1 counter dnat to 127.0.0.1 comment mirrored;
```

both are methods to prevent the PREROUTING NAT rule.
there is no change in behavior when using with Docker, but i think this is safer because it does not affect the chain managed by iptables.

if you like, please try this as well.

**Comment by andreaprilucm** [comment]
@felipecrs  When you disable iptables manipulations by docker, you can't have isolated networks for your docker compose stacks.
It has the same effect as adding `--network host` to each container. This is acceptable for simple use cases where you start a few unrelated containers.

Imagine that I have an application composed of two containers: a web server (Tomcat port 8080) and a database (oracle port 1521). I create a `compose.yaml` file containing two services : `tomcat `and `oracle`. And I start the pair of containers with `docker compose up`.
 In my `compose.yaml`, I only expose the web server port to the outside world. The application running in the Tomcat container refers to the database as "oracle port 1521" (`oracle `is the name of the service in compose.yaml).

When iptables manipulation is disabled, the Oracle database binds to 127.0.0.1:1521. If you try to start a second compose stack, Oracle will complain that the port is already in use. 

If iptables manipulation is enabled (default), each stack will have its own network ; e.g. "172.19.0.0/16" and "172.20.0.0/16" and Oracle in the first stack will bind to 172.19.0.3:1521 wile the one in the second stack will bind to 172.20.0.3:1521. In the tomcat container of the first stack, the `oracle` host name refers to 172.19.0.3 while in the Tomcat container the `oracle` host name refers to 172.20.0.3.

The fix proposed by Shigenobu NAT the packets arriving on the loopback0 interface (the one the Windows packets come from) before forwarding them on the docker0 interface (the one Docker bridges its networks). See his comment of Sep, 30 2023 for the explanation of the problem.

If you want more details on the networking of Docker, see https://docs.docker.com/network/network-tutorial-standalone/ and https://docs.docker.com/network/network-tutorial-host/.

So, disabling iptables manipulation is fine for some simple use cases but you will quickly get stuk.

**Comment by felipecrs** [comment]
Thanks both, very useful information.

**Comment by felipecrs** [comment]
Although `networkingMode=mirrored` is great, I hope it does not get promoted to default without having this issue resolved first. 😅

**Comment by felipecrs** [comment]
Just to renew this issue, I tested again with the just released WSL 2.2.1, but same problem still occurs.

**Comment by NiklasBr** [comment]
Still an issue with v4.28.0

**Comment by jsayer101** [comment]
Hello,

We're trying to have a setup of WSL working behind a corporate VPN (Cisco Any Connect).

On Win11/WSL2, so far we have to use networkingMode=mirrored to get WSL to be able to access the Internet. (works fine)

But we're also trying to use docker inside WSL. Now, with this setup we can't access our containers that exposes specific ports from our Windows machine. For example portainer on port 9000.

This issue cause us to not be able to use Docker on WSL behind a corporate VPN.

```
[wsl2]
vmIdleTimeout=-1
autoProxy=false
networkingMode=mirrored
dnsTunneling=true
```

```
Version WSL : 2.1.5.0
Version du noyau : 5.15.146.1-2
Version WSLg : 1.0.60
Version MSRDC : 1.2.5105
Version direct3D : 1.611.1-81528511
Version de DXCore : 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Version de Windows : 10.0.22621.3296
```

If there is any news on this issue I'll be greatful to hear from you guys.

watching :eyes:

**Comment by foureight84** [comment]
> Hello,
> 
> We're trying to have a setup of WSL working behind a corporate VPN (Cisco Any Connect).
> 
> On Win11/WSL2, so far we have to use networkingMode=mirrored to get WSL to be able to access the Internet. (works fine)
> 
> But we're also trying to use docker inside WSL. Now, with this setup we can't access our containers that exposes specific ports from our Windows machine. For example portainer on port 9000.
> 
> This issue cause us to not be able to use Docker on WSL behind a corporate VPN.

> @jsayer101 I don't understand your configuration. Where is your portainer server deployed, your portainers agents, your firewal/VPN server, the browser you use to access your portainer server (outside your corporate LAN?)... ? Could you draw a picture of it?
> 
> If you want to access ports on your WSL from other hosts, does your windows firewall authorize it?
> 
> Have you tried the startup script propose above?


Yea something is out of order. Corporate firewall wouldn't prevent you from accessing WSL on your own computer.

Anyway, add this to your WSL OS (https://github.com/microsoft/WSL/issues/10494#issuecomment-1951848642) and add the following settings to your .wslconfig. It's similar to what you have but with additional flags. This is working fine across all of my Win11 PCs

```
[wsl2]
networkingMode=mirrored
firewall=false
dnsTunneling=true

[experimental]
useWindowsDnsCache=true
hostAddressLoopback=true
```

**Comment by dharapvj** [comment]
@shigenobuokamoto 

>could you please try this?  
https://gist.github.com/shigenobuokamoto/b565d468541fc8be7d7d76a0434496a0

Your hack works fine for me. Thank you so much!.. Just one small issue.. it breaks other IPv6 access going to internet for me. My docker is not configured for IPv6. So.. could you help me understand what parts of the gist should be removed so that it does not mess with ipv6 routing stuff?

```
❯ ssh -6 ubuntu@xxx.com -vv
OpenSSH_8.9p1 Ubuntu-3ubuntu0.6, OpenSSL 3.0.2 15 Mar 2022
debug1: Reading configuration data /home/vj/.ssh/config
debug1: /home/vj/.ssh/config line 1: Applying options for *
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: include /etc/ssh/ssh_config.d/*.conf matched no files
debug1: /etc/ssh/ssh_config line 21: Applying options for *
debug2: resolving "xxx.com" port 22
debug1: Connecting to xxx.com [2406:0000:000:0000:000:0000:0000:c711] port 22.
debug1: connect to address 2406:0000:000:0000:000:0000:0000:c711 port 22: Connection timed out
ssh: connect to host xxx.com port 22: Connection timed out
```

With normal `mirrored` mode, this connectivity used to work. I can still connect fine via same command from windows host. just not from WSL.

**Comment by shigenobuokamoto** [comment]
@dharapvj 
> Your hack works fine for me. Thank you so much!.. Just one small issue.. it breaks other IPv6 access going to internet for me. My docker is not configured for IPv6. So.. could you help me understand what parts of the gist should be \
removed so that it does not mess with ipv6 routing stuff?

if you exclude the line containing "ip6", IPv6 support will be disabled.

```
$ grep -v ip6 network-mirrored.service > mirrored-noip6.service
```

however, since this support controls the source port for sources other than WSL container (= the traffic routed by WSL container), it is strange that access via SSH is no longer possible.

**Comment by hyp530** [comment]
Waiting for WSL to fix this issue.

**Comment by impzero** [comment]
same issue

**Comment by AlencarGabriel** [comment]
same issue, please MS, help us!

**Comment by keith-horton** [comment]
Hi there.

We have a few questions as we narrow down what is going on.

The below is referencing the troubleshooting docs in https://github.com/MicrosoftDocs/WSL/blob/main/WSL/troubleshooting.md

Sadly not all changes have been replicated yet to https://learn.microsoft.com/en-us/windows/wsl/troubleshooting :(


1. can folks who are hitting this confirm they do not have a separate Network Manager running?
see "Docker container issues when its Network Manager is running" in the above troubleshooting article.

2. have folks who have hit this looked at the troubleshooting article we wrote? Can you give us feedback if that works for you, or if that cannot be done for a reason with your scenario?
see: "Docker container issues in WSL2 with Mirrored networking mode enabled when running under the default networking namespace"


Thanks for your help!
We're trying to understand exactly what is going on.

**Comment by AlencarGabriel** [comment]
> Hi there.
> 
> We have a few questions as we narrow down what is going on.
> 
> The below is referencing the troubleshooting docs in https://github.com/MicrosoftDocs/WSL/blob/main/WSL/troubleshooting.md
> 
> Sadly not all changes have been replicated yet to https://learn.microsoft.com/en-us/windows/wsl/troubleshooting :(
> 
> 1. can folks who are hitting this confirm they do not have a separate Network Manager running?
>    see "Docker container issues when its Network Manager is running" in the above troubleshooting article.
> 2. have folks who have hit this looked at the troubleshooting article we wrote? Can you give us feedback if that works for you, or if that cannot be done for a reason with your scenario?
>    see: "Docker container issues in WSL2 with Mirrored networking mode enabled when running under the default networking namespace"
> 
> Thanks for your help! We're trying to understand exactly what is going on.

In my scenario, I don't have `network-manager.service`:

![image](https://github.com/microsoft/WSL/assets/10109480/89cf8df9-a256-4fe4-99d5-0126b7f0dfb8)

I use docker-ce, not docker-desktop. My case:

From WSL:  `curl localhost:8080` OK (service 8080 running in windows host)

From container docker: `curl localhost:8080` = `Couldn't connect to server` - OK localhost here is the container loopback.

From container docker: `curl 10.173.16.72:8080` = `Couldn't connect to server`.  - 10.173.16.72 is IP of WSL and Windows Host.

**Comment by shigenobuokamoto** [comment]
seems like a new issue that is different from the current flow.

> From container docker: curl 10.173.16.72:8080 = Couldn't connect to server. - 10.173.16.72 is IP of WSL and Windows Host.

this is the `WSL`  ip address from the container's perspective, not `WSL and Windows Host`.

then, i tried to think of a way to implement this method.
with these steps, you can connect 10.173.16.72:8080 from the container to 8080 on the Windows host.

1. enable the target port to overlap with Windows Host

.wslconfig
```
[experimental]
ignoredPorts=8080
```

2. 10.173.16.72:8080 -> 127.0.0.1:8080 with socat

```
$ sudo socat TCP-LISTEN:8080,bind=10.173.16.72,reuseaddr,fork TCP:127.0.0.1:8080
```

**Comment by AlencarGabriel** [comment]
> seems like a new issue that is different from the current flow.
> 
> > From container docker: curl 10.173.16.72:8080 = Couldn't connect to server. - 10.173.16.72 is IP of WSL and Windows Host.
> 
> this is the `WSL` ip address from the container's perspective, not `WSL and Windows Host`.
> 
> then, i tried to think of a way to implement this method. with these steps, you can connect 10.173.16.72:8080 from the container to 8080 on the Windows host.
> 
> 1. enable the target port to overlap with Windows Host
> 
> .wslconfig
> 
> ```
> [experimental]
> ignoredPorts=8080
> ```
> 
> 2. 10.173.16.72:8080 -> 127.0.0.1:8080 with socat
> 
> ```
> $ sudo socat TCP-LISTEN:8080,bind=10.173.16.72,reuseaddr,fork TCP:127.0.0.1:8080
> ```

Sorry @shigenobuokamoto , I mixed up the problems.

> Regarding the `10.173.16.72` ip, it's from WSL and Windows Host. It's different because it's the IP when I'm connected to the VPN. I also tried the IP of the Wi-fi network interface but to no avail.

But I had the same port forwarding problem with my docker-ce and `networkingMode=mirrored`.

`docker run --rm -it -p 8082:8082 alpine nc -p 8082 -l`

Trying to connect via the browser or curl on the Windows machine doesn't work (http://localhost:8082). However, if I connect via WSL, it works.

I set `{"iptables": false}` in daemon.json and it worked. Without the iptables modification it only worked with `--network-host`.

Now I think we're talking about the same thing in this issue.

**Comment by shigenobuokamoto** [comment]
@AlencarGabriel, sorry if I offended you

mirrored network has a rather special structure. it would be nice to have an official document, but i have written my thoughts below, so please use them as a reference.

https://github.com/microsoft/WSL/discussions/11326#discussioncomment-8874149

for connections from Windows Host to docker on WSL, you can control traffic to the container using daemon.json or nftables. currently, it is not possible to create the same behavior as native linux without controlling it.

10.173.16.72 is its own address to linux. since it processes the traffic that comes here by itself, it rejects connect to 8080 that is not open. but Linux itself can connect to 10.173.16.72:8080 (on Windows Host). this is where the magic of mirrored network.

this issue and similar docker-related issues pointed to connect from Windows host to container service.
in that sense, i think that connecting from container to Windows poses a new issue. and a different (ad-hoc) solution is required than what has been shown so far.

**Comment by watchingfun** [comment]
> Interestingly, the port forwarding does work from another machine on the same network as host. Just not on the host machine itself.
> 
> Machine A:
> 
> * Change networkingMode to mirrored
> * docker run -d -p 8080:80 nginx:alpine
> * Go to localhost:8080
> * Cannot connect
> * Set Windows firewall rules to allow 8080 inbound
> * Go to localhost:8080
> * Cannot connect
> 
> Machine B:
> 
> * Go to A:8080
> * works

🤣Damn, I've been messing around all afternoon. I turned off the firewall and forwarded the ports, but I just don't know why the telnet connection won't connect. Later I saw what you said and found out that it doesn’t work locally.

**Comment by AntonOfTheWoods** [comment]
I have the following in my `.wslconfig`

```
[wsl2]
memory=20GB

networkingMode=mirrored
dnsTunneling=true
firewall=true
autoProxy=true

[experimental]
sparseVhd=true
# requires dnsTunneling
bestEffortDnsParsing=true
useWindowsDnsCache=true
```

I have a `wireguard` VPN running on my Windows host and I have no extra config in my wsl or docker, except in my `/etc/sysctl.conf` (because certain apt repos are utterly unusable using ipv6, at least over my Wireguard vpn):

```
net.ipv6.conf.all.disable_ipv6=1
net.ipv6.conf.default.disable_ipv6=1
net.ipv6.conf.lo.disable_ipv6=1
```

I have no extra Windows Firewall config and I have absolutely no issues accessing the web from my host or `wsl`, and `docker` behaves exactly how I would expect it too (I can access everything from both the host and `wsl`). It works perfectly whether I turn the VPN on or off.

I am using vanilla `docker` installed from the instructions https://docs.docker.com/engine/install/ubuntu/. This also works perfectly with the new Ubuntu 24.04 LTS (Ubuntu Preview in the Store).

**Comment by felipecrs** [comment]
What is your distribution? I'm tried configuring my WSL like yours, but I still can reproduce the issue.

https://github.com/microsoft/WSL/assets/29582865/38338c55-979d-4a25-a7d5-57b57497b343

```console
❯ wsl --version
WSL version: 2.2.3.0
Kernel version: 5.15.153.1-2
WSLg version: 1.0.61
MSRDC version: 1.2.5326
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.26091.1-240325-1447.ge-release
Windows version: 10.0.22631.3447
```

**Comment by felipecrs** [comment]
BTW I can confirm that @shigenobuokamoto's [workaround](https://gist.github.com/shigenobuokamoto/b565d468541fc8be7d7d76a0434496a0) works like a charm.

**Comment by keith-horton** [comment]
To follow-up on this thread, we are actively working to address this. We are considering solution such as that shown by @shigenobuokamoto  earlier in this thread. We are currently considering options and possible side-effects for the multitude of ways Linux might be configured by users.

Thanks everyone for the detailed feedback!

**Comment by underlines** [comment]
Wow, I couldn't get it to work in mirrored mode for 4 hours, then it randomly worked after restarting services, killing processes, restarting wsl, running docker desktop as admin etc.
The next day I couldn't open docker desktop, the engine would just not start, for 2 hours.

Modifying my config finally worked:

[wsl2]


I added:
```
firewall=false
dnsTunneling=true
```
so my config now looks like this:
```
networkingMode=mirrored
# localhostForwarding=true
firewall=false
dnsTunneling=true

[experimental]
autoMemoryReclaim=gradual
sparseVhd=true
```

**Comment by SiggiSmara** [comment]
I'm using docker-ce, behind a corporate proxy/VPN with mirrored networking (no other wsl settings) and the only thing that has worked so far that allows access to docker containers ports from the windows host side is to not use iptables on the dockerd side.

**Comment by tyeth** [comment]
Just got this running recently with @underlines config.

Forgot that mirrored mode requires a signed in user, which fails if offline (even signed in and online before reboot), which is beyond frustrating, and will cause me to recommend against docker desktop version of docker.

**Comment by spyro2000** [comment]
Network Mode mirrored does not work for me with docker. Looks like port 2375 is only reachable via IPv6 (`netstat - nl | grep 2375` only returns a mapping for 'tcp6').

Why I can reach http://localhost:2375/version via Browser, e. g. Testcontainers does not find a docker environment (despite DOCKER_HOST set to 'tcp://localhost:2375').

Even after 8 years, docker is just a nightmare with WSL(2).I am literally the only dev using Windows anymore on my team, everyone else is using Macs now where it just works.

**Comment by flanger001** [comment]
I keep coming back to this and @shigenobuokamoto's solution is the only thing that works for me.

**Comment by mwwhited** [comment]
After updating WSL I'm running into the same issue that `networkingNode=mirrored` no longer works.  If I set it to NAT I can get to the WSL/docker container from the host machine but I can no longer get to it from the network at all.

**Comment by jweaston** [comment]
This issue seems to have been resolved for me with WSL 2.2.4.0. Can anyone else confirm this so we can close the issue?

**Comment by mrcaidev** [comment]
> This issue seems to have been resolved for me with WSL 2.2.4.0. Can anyone else confirm this so we can close the issue?

No. I still need to disable `iptables` to make it work.

```
WSL version: 2.2.4.0
Kernel version: 5.15.153.1-2
WSLg version: 1.0.61
MSRDC version: 1.2.5326
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.26091.1-240325-1447.ge-release
Windows version: 10.0.22631.3737
```

**Comment by CatalinFetoiu** [comment]
hello, thanks for your patience while we investigated this

we are talking with Docker to address this in their product, but in the meantime, you can run these commands to update the Docker nftables rules to allow localhost port forwarding to work

step 1) Run "sudo nft -a list chain ip nat DOCKER". Example output shown below.
table ip nat {
        chain DOCKER { # handle 4
                iifname "docker0" counter packets 0 bytes 0 return # handle 7
                iifname != "docker0" meta l4proto tcp tcp dport 8080 counter packets 0 bytes 0 dnat to 172.17.0.2:80 # handle 12
        }
}

step 2)
using the handle of the second rule in the DOCKER chain from step 1 (the rule that includes the "dnat" keyword), run the following command:
(please make sure to use the handle from your output in step 1, as the handle number can change)

sudo nft -a delete rule ip nat DOCKER handle 12

step 3) Run the following command. Make sure that the ports and IP address match your output from step 1

sudo nft add rule ip nat DOCKER iifname != "docker0" iifname != "loopback0" meta l4proto tcp tcp dport 8080 counter dnat to 172.17.0.2:80

**Comment by jweaston** [comment]
> This issue seems to have been resolved for me with WSL 2.2.4.0. Can anyone else confirm this so we can close the issue?

I guess I should also specify I am not using docker desktop. I installed Docker directly in WSL.

**Comment by shigenobuokamoto** [comment]
@CatalinFetoiu 
it may not be practical to modify the nft rules set by Docker every time.
instead, if you run the following two lines, traffic from the Windows host will be handled by 127.0.0.1 (= docker-proxy).


```
sudo nft  add chain   'ip nat WSLPREROUTING { type nat hook prerouting priority dstnat - 1; policy accept; }'
sudo nft  insert rule 'ip nat WSLPREROUTING iif loopback0 ip daddr 127.0.0.1 counter dnat to 127.0.0.1'
```

i have been using this rule for a long time and it does not seem to have any side effects.

**Comment by AntonOfTheWoods** [comment]
@shigenobuokamoto's solution continues to be rock-solid over many months of 10h/day usage. Write yourself a personal VM setup guide and just copy paste their stuff.

**Comment by mwwhited** [comment]
I thought I was only having this issue with docker but its really any wsl instance.  I can connect between WSL instances but windows to wsl does not work.  I am running in mirrored mode.  This works fine on 2 of the 3 computers I have available to me.  

I've tried a clean reinstall of wsl on this problem machine but nothing works. there is no configuration/version difference between these machines.

**Comment by mwwhited** [comment]
I have no idea how to explain this one... but I have mirrored mode working again after playing around with wireshark in an effort to see where the TCP/IP stack was breaking down.  after resetting promiscuous mode on all of the network devices the windows to wsl communications started working again.  This was after multiple reinstalls of wsl, hyper-v, docker as well as all of the different firewall scripts in the related issues and resetting the ip stack in windows.  My best guess is when I switched from cabled ethernet to my wifi enabled hotspot after a network outage something in the network stack was flipped breaking the communications between windows and wsl.  But after toggling promiscuous mode on all of the devices somehow it was set back to the manner required to make "mirrored" work correctly.  

It would be nice to have more information but I have no idea how to reproduce the breaking issue or the fix.  I was actually rather annoyed when I found the stack was working again as I did not make any intentional changes.  I was just playing in wireshark trying to find the `curl http://localhost` with it timing out... just to try it again and have it start returning results.  

Good luck to the person that gets to hunt down the issue in the TCP/IP stack.

**Comment by mwwhited** [comment]
More interestingly... it only works when I have wireshark doing a local loopback capture.

**Comment by itispip** [comment]
> @CatalinFetoiu it may not be practical to modify the nft rules set by Docker every time. instead, if you run the following two lines, traffic from the Windows host will be handled by 127.0.0.1 (= docker-proxy).
> 
> ```
> sudo nft  add chain   'ip nat WSLPREROUTING { type nat hook prerouting priority dstnat - 1; policy accept; }'
> sudo nft  insert rule 'ip nat WSLPREROUTING iif loopback0 ip daddr 127.0.0.1 counter dnat to 127.0.0.1'
> ```
> 
> i have been using this rule for a long time and it does not seem to have any side effects.

seems still have side efforts:
1. win host --> service in wsl, ok
2. wsl --> service in win host, ok
3. win host --> service in wsl docker, ok through localhost/127.0.0.1
4. wsl docker --> service in win host, no

above bullet 4 in WSL2 NAT workable.

**Comment by shigenobuokamoto** [comment]
@itispip
> 4. wsl docker --> service in win host, no

https://github.com/microsoft/WSL/issues/10494#issuecomment-2050817875

this is the way I thought about it before.
it seems difficult to handle this with nftables, so i am using socat to relay traffic.

**Comment by craigloewen-msft** [comment]
Hi folks, we believe we've found a fix for this that needs to be added to the Docker engine. We've filed a bug for it and are working with the docker team to investigate adding it in: https://github.com/moby/moby/issues/48056

**Comment by mwwhited** [comment]
> Hi folks, we believe we've found a fix for this that needs to be added to the Docker engine. We've filed a bug for it and are working with the docker team to investigate adding it in: [moby/moby#48056](https://github.com/moby/moby/issues/48056)

As a note, this is not just docker.  Even without docker and resetting WSL entirely this still occurs. mirrored mode does not work correctly with regular WSL images such as Ubuntu.

**Comment by mwwhited** [comment]
There are several here all describing the same issue.  Opening more is not going to fix it any better.

**Comment by felipecrs** [comment]
But I can tell you this issue will very likely be closed as soon as the Docker CE issue is resolved.

**Comment by mwwhited** [comment]
As already noted above I have cross mentioned the issue to try to get an actual fix instead of just a pile of work arounds.  
https://github.com/microsoft/WSL/issues/11600

**Comment by jweaston** [comment]
> Reference

Well it looks like the issue wasn't actually fixed for me. Not sure what I did or what changed to fix and then unfix the issue. 

Also when running the `sudo nft -a list chain ip nat DOCKER` I get a new rule for each port I forward in the container.

# Warning: table ip nat is managed by iptables-nft, do not touch!
table ip nat {
        chain DOCKER { # handle 10
                iifname "docker0" counter packets 0 bytes 0 return # handle 29
                iifname "br-c360e1083765" counter packets 65 bytes 3981 return # handle 13
                iifname != "br-c360e1083765" tcp dport 8085 counter packets 0 bytes 0 dnat to 172.19.0.2:54002 # handle 30
                iifname != "br-c360e1083765" tcp dport 10001 counter packets 0 bytes 0 dnat to 172.19.0.2:10000 # handle 32
                iifname != "br-c360e1083765" tcp dport 9001 counter packets 0 bytes 0 dnat to 172.19.0.2:9001 # handle 34
                iifname != "br-c360e1083765" udp dport 1024 counter packets 0 bytes 0 dnat to 172.19.0.2:1024 # handle 36
        }
}

Where it looks like the `c360e1083765` is the ID of the container. Not sure if this is different because I'm using the docker engine over docker desktop.

**Comment by shigenobuokamoto** [comment]
although this issue occurs with Docker, i believe it is a compatibility issue with WSL.
the following parameters set by mirrored networking are unexpected from Docker.

```
net.ipv4.conf.loopback0.route_localnet = 1
```
my nftables rules partially cancels the effect of this parameter.
Docker can run even with the above parameters set, but we will encounter another issue where the Windows host cannot route to the networks added inside WSL.

if the Windows host could reflect the routing to the network inside WSL, the fourth point raised by @itispip would be easily solved.

**Comment by felipecrs** [comment]
I just tested WSL 2.3.11 and unfortunately the problem remains there.

**Comment by keith-horton** [comment]
@felipecrs : right - 2.3.1.1 fixes a variety of Linux container scenarios. This in particular will require Docker to change their route - tracked here: https://github.com/moby/moby/issues/48056.

We are in contact with them and the fix should hopefully be released soon.

**Comment by shigenobuokamoto** [comment]
@keith-horton 
(https://github.com/microsoft/WSL/releases/tag/2.3.11)[WSL 2.3.11]
> Update mirrored mode loopback routing to support connections between Windows host and Linux Docker/Podman containers (solves #11136,#11468,#11758,#10926)

this improvement gives us more options.

in the daemon.json

> "userland-proxy": false

this now works as well.

but

https://github.com/moby/moby/pull/48075

this change is made, 'userland-proxy: false' may no longer work.
i think it needs a little adjustment.

**Comment by shigenobuokamoto** [comment]
this may seem unnecessary since moby takes care of it, but this is a new improvement plan.

WSL 2.3.11 brings several improvements to communication with Windows host.

therefore, you can now use the following rules to improve communication from Windows host to Docker containers:

```
$ nft add rule ip nat WSLPOSTROUTING iif "loopback0" ip saddr 127.0.0.0/8 ip daddr != 127.0.0.0/8 counter masquerade
$ sysctl -w net.ipv4.conf.all.route_localnet=1
```

**Comment by codeart1st** [comment]
@shigenobuokamoto with WSL pre-release 2.3.11 would you still recommend your systemd script workaround?

https://gist.github.com/shigenobuokamoto/b565d468541fc8be7d7d76a0434496a0

**Comment by felipecrs** [comment]
> @shigenobuokamoto with WSL pre-release 2.3.11 would you still recommend your systemd script workaround?
> 
> https://gist.github.com/shigenobuokamoto/b565d468541fc8be7d7d76a0434496a0

As an user, I can it still works fine.

**Comment by shigenobuokamoto** [comment]
@codeart1st 
﻿work on this issue is ongoing in moby (=Docker Engine), so until that is released you can use my script to mitigate the issue.
essentially, the issue is that network within WSL behaves differently from normal Linux, so i think it would be best to deal with it on the WSL side.

WSL 2.3.11 includes some improvements to mirrored networking, so i am looking into how to use them to improve the experience. this is the prototype new network-mirrored.service. please try this too.

https://gist.github.com/shigenobuokamoto/540c5f09a03eb07149501e99a6c8d82b

- taking advantage of the fact that connect from Docker container network to Windows host is now possible, i wrote a more natural nftables.
- added support for communication between Windows and Linux using 127.0.0.0/8, so i worte the missing route.

**Comment by felipecrs** [comment]
> the issue is that network within WSL behaves differently from normal Linux, so i think it would be best to deal with it on the WSL side.

Also, this would mean that whatever application that relies on the same feature as docker would have the same issue. It could be an old application that doesn't receive more support for example. Or some application that refuses to implement a fix just for WSL.

So, yeah, I totally agree. The ideal would have been the fix to happen on WSL side.

**Comment by Parsifa1** [comment]
When will this fix be built into wsl?Now we can only use service script to make it work, I don't think this is a long term solution.

Besides, I use nixos-wsl, this is my example module that can be used as a reference for nixos user, from @shigenobuokamoto
```nix

{pkgs, ...}: {
    systemd.services.network-mirrored = {
      description = "network-mirrored";
      enable = true;
      wants = ["network-pre.target"];
      wantedBy = ["multi-user.target"];
      before = ["network-pre.target" "shutdown.target"];
      serviceConfig = {
        User = "root";
        ExecStart = [
          ''
            /bin/sh -ec '\
            [ -x /usr/bin/wslinfo ] && [ "$(/usr/bin/wslinfo --networking-mode)" = "mirrored" ] || exit 0;\
            echo "\
            add chain   ip nat WSLPREROUTING { type nat hook prerouting priority dstnat - 1; policy accept; };\
            insert rule ip nat WSLPREROUTING iif loopback0  ip daddr 127.0.0.1 counter dnat to 127.0.0.1 comment mirrored;\
            "|${pkgs.nftables}/bin/nft -f -\
            '
          ''
        ];

        ExecStop = [
          ''
            /bin/sh -ec '\
              [ -x /usr/bin/wslinfo ] && [ "$(/usr/bin/wslinfo --networking-mode)" = "mirrored" ] || exit 0;\
              for chain in "ip nat WSLPREROUTING";\
              do\
                handle=$(${pkgs.nftables}/bin/nft -a list chain $chain | sed -En "s/^.*comment \\"mirrored\\" # handle ([0-9]+)$/\\1/p");\
                for n in $handle; do echo "delete rule $chain handle $n"; done;\
              done|${pkgs.nftables}/bin/nft -f -\
            '
          ''
        ];
        RemainAfterExit = "yes";
      };
    };
}
```

**Comment by zaaack** [comment]
here is a really dirty solution, I just wrote a nodejs tcp reverse proxy server to map the docker port to another port, so we can using another port because its created by a non-docker process.

```js
import net from 'net'
let map = {
  15432: 5432,
  13000: 3000,
}
for (const from in map) {
  const to = map[from]
  net
    .createServer((server) => {
      const client = net.createConnection(to)
      server.pipe(client)
      client.pipe(server)
    })
    .listen(from)
    .on('listening', () => {
      console.log(`tcp-proxy: ${from} -> ${to}`)
    })
    .on('error', (err) => {
      console.error(err)
    })
}

```

**Comment by felipecrs** [comment]
@zaaack, if you are looking for a temporary solution, this one is the best you'll find: https://github.com/microsoft/WSL/issues/10494#issuecomment-2241498340

**Comment by zaaack** [comment]
> @zaaack, if you are looking for a temporary solution, this one is the best you'll find: [#10494 (comment)](https://github.com/microsoft/WSL/issues/10494#issuecomment-2241498340)

thanks, but my wsl  is older and doesnt set systemd

**Comment by dannyhpy** [comment]
> thanks, but my wsl is older and doesnt set systemd

You can use `iptables -t nat -I DOCKER -i loopback0 -d 127.0.0.0/8 -j RETURN`[¹](https://github.com/moby/moby/pull/48075/files#diff-aa946998315144e84ceffe7bd84918fd8d34183af1cae444ce90e2b620ec8371R581) or any of the variants[²](https://github.com/microsoft/WSL/issues/10494#issuecomment-2240981147) that have been mentioned above.

**Comment by jweaston** [comment]
Has anyone had success configuring WSL to allow connections from containers in WSL to the windows host? 
I made some attempts on my own to forward used ports, 5966 in this case, from WSL to the windows host through the nft but was unsuccesfull.
Maybe some one can tell me what I did wrong.

`sudo nft add chain ip nat prerouting { type nat hook prerouting priority 0 \; }
sudo nft add chain ip nat postrouting { type nat hook postrouting priority 100 \; }
sudo nft add rule ip nat prerouting tcp dport 5966 counter dnat to 127.0.0.1:5966
sudo nft add rule ip nat postrouting ip saddr 172.0.0.0/8 counter masquerade`

**Comment by shigenobuokamoto** [comment]
@jweaston 
you can access from Docker container in WSL to Windows **host address**, WSL 2.3.11 and later.

.wslconfig
```
[experimental]
hostAddressLoopback=true
```

```
Windows host                  WSL Linux                         Docker Container
127.0.0.1    --- loopback --- 127.0.0.1   --- not connected --- 127.0.0.1(same address but unrelated)
192.168.1.10 --- loopback --- 192.168.1.10
                                  | routed
     (unknown network)        172.17.0.1  ----- connected ----- 172.17.0.2
```

docker container (in WSL) can connect to 192.168.1.10:5966.
if you try connect to 127.0.0.1:5966, the packet is sent to container's 127.0.0.1. so cannot connect to windows host.

**Comment by CatalinFetoiu** [comment]
hello. Docker completed the PR that fixes this issue - https://github.com/moby/moby/pull/48514
This was included in the following Docker release - https://github.com/moby/moby/releases/tag/v27.3.0

Please try Docker version 27.3.0

**Comment by felipecrs** [comment]
It works in my test. :)

**Comment by unclemusclez** [comment]
My favorite flavor of Linux is Ubuntu via Docker via Ubuntu via WSL via Windows 11.

**Comment by BolteDev** [comment]
Just to add to this...

I recently set up Ubuntu 24.04.1 with Docker and it works fine in Mirrored mode (Docker 27.3.1)
But on my Alpine install (Docker 26.1.5) it refuses to work unless I set it back to NAT.

Was driving me crazy

**Comment by unclemusclez** [comment]
> Just to add to this...
> 
> I recently set up Ubuntu 24.04.1 with Docker and it works fine in Mirrored mode (Docker 27.3.1) But on my Alpine install (Docker 26.1.5) it refuses to work unless I set it back to NAT.
> 
> Was driving me crazy
> 
> EDIT: Does anyone have any advice on manually installing latest version in Alpine without using Dockers setup script that tries to force mt to use Docker Desktop?

can i see all of your relevant settings because i still cant get it to work. I am trying to run a simple gradio app inside of a docker instance.

**Comment by unclemusclez** [comment]
> Just to add to this...
> 
> I recently set up Ubuntu 24.04.1 with Docker and it works fine in Mirrored mode (Docker 27.3.1) But on my Alpine install (Docker 26.1.5) it refuses to work unless I set it back to NAT.
> 
> Was driving me crazy
> 
> EDIT: Does anyone have any advice on manually installing latest version in Alpine without using Dockers setup script that tries to force mt to use Docker Desktop?

@BolteDev can i see all of your relevant settings because i still cant get it to work. I am trying to run a simple gradio app inside of a docker instance. this used to work fine and since i've started trouble shooting everything i have no idea what are the proper settings.

**Comment by BolteDev** [comment]
> > Just to add to this...
> > I recently set up Ubuntu 24.04.1 with Docker and it works fine in Mirrored mode (Docker 27.3.1) But on my Alpine install (Docker 26.1.5) it refuses to work unless I set it back to NAT.
> > Was driving me crazy
> > EDIT: Does anyone have any advice on manually installing latest version in Alpine without using Dockers setup script that tries to force mt to use Docker Desktop?
> 
> @BolteDev can i see all of your relevant settings because i still cant get it to work. I am trying to run a simple gradio app inside of a docker instance. this used to work fine and since i've started trouble shooting everything i have no idea what are the proper settings.

---------

I haven't done anything different than following normal installations and setting up docker, other than my wslconfig having networkingMode=mirrored.

I was setting up an Alpine install and nginx just wasn't working until I disabled it and went back to regular NAT along with localhostForwarding=true, I thought the docker just wasn't working until I did a curl against localhost from within the instance and realised it was working, which lead me here.

I'm still pretty new to Linux side of things, but I would say if you are lost with it, start from scratch. I would say try and get something simple like Nginx working then at least you know the networking side works and it will then be a docker image issue or not.

But basically all I did was install openrc (I'm using the 3mb rootfs Alpine) do the basic setup to start it in /etc/init-wsl:

```bash
# OpenRC
mount --make-rshared /
mkdir -p /run/openrc
touch /run/openrc/softlevel
sysctl net.ipv4.ip_unprivileged_port_start=80 > /dev/null 2>&1

openrc
rc-service networkmanager start
rc-service docker start
```

Then ran the standard nginx image as a test on 80:80

**Comment by keith-horton** [comment]
If you're seeing issues, please be sure you have the latest Docker + latest WSL.
Some applications may resolve the name of the local device to an assigned address - instead of using loopback. In those cases you'll want to set [experimental] hostAddressLoopback=true.

https://learn.microsoft.com/en-us/windows/wsl/wsl-config 

hostAddressLoopback (bool - default == false)
Only applicable when wsl2.networkingMode is set to mirrored. When set to True, will allow the Container to connect to the Host, or the Host to connect to the Container, by an IP address that's assigned to the Host. The 127.0.0.1 loopback address can always be used,this option allows for all additionally assigned local IP addresses to be used as well. Only IPv4 addresses assigned to the host are supported.

**Comment by unclemusclez** [comment]
> If you're seeing issues, please be sure you have the latest Docker + latest WSL. Some applications may resolve the name of the local device to an assigned address - instead of using loopback. In those cases you'll want to set [experimental] hostAddressLoopback=true.
> 
> https://learn.microsoft.com/en-us/windows/wsl/wsl-config
> 
> hostAddressLoopback (bool - default == false) Only applicable when wsl2.networkingMode is set to mirrored. When set to True, will allow the Container to connect to the Host, or the Host to connect to the Container, by an IP address that's assigned to the Host. The 127.0.0.1 loopback address can always be used,this option allows for all additionally assigned local IP addresses to be used as well. Only IPv4 addresses assigned to the host are supported.

@keith-horton so i want them both to true? i run other apps withing WSL that work fine, but when i try to host the docker containers from the run command inside of WSL it does not connect to a forwarded port.

**Comment by keith-horton** [comment]
Right: some apps may want to connect to an assigned address instead of loopback. In those cases, it might help those solutions end-to-end to *also* enable hostAddressLoopback

(hostAddressLoopback is only applicable to Mirrored mode)

**Comment by unclemusclez** [comment]
> Right: some apps may want to connect to an assigned address instead of loopback. In those cases, it might help those solutions end-to-end to _also_ enable hostAddressLoopback
> 
> (hostAddressLoopback is only applicable to Mirrored mode)

@keith-horton brother, you just changed my life. thank you

**Comment by codeart1st** [comment]
```
docker --version
Docker version 27.3.1, build ce12230

wsl.exe --version
WSL version: 2.3.14.0
Kernel version: 6.6.36.3-1
WSLg version: 1.0.64
MSRDC version: 1.2.5326
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.26100.1-240331-1435.ge-release
Windows version: 10.0.22631.4169
```

Disabled and stopped the systemd service helper for https://gist.github.com/shigenobuokamoto/540c5f09a03eb07149501e99a6c8d82b.

Propably most things work, but for no real reason I can't connect with my Oracle DB Container anymore. Reenabled the systemd service helper and my Oracle DB Container works again.

**Comment by CatalinFetoiu** [comment]
closing since the issue is resolved
if you encounter the issue after using Docker version 27.3.0 or later, please open a separate issue

**Comment by cacayue** [comment]
I switched to using 127.0.0.1 instead of localhost, and now the connection works as expected.
Reference: [Docker in WSL – Can’t connect to SQL Server on ‘localhost’ but with ‘127.0.0.1’](https://sourcebae.com/blog/docker-in-wsl-cant-connect-to-sql-server-on-localhost-but-with-127-0-0-1-sql-server-2017-latest/)

**Comment by felipecrs** [comment]
This issue is fixed. Are you using an old version of docker by any chance?

**Comment by Dokman** [comment]
i'm getting this error too
